src/ast.rs

```diff
@@ -0,0 +1,364 @@
+
+/// A `node` statement that creates a new graph node
+#[derive(Debug, Eq, PartialEq)]
+pub struct CreateGraphNode {
+    pub node: Variable,
+}
+
+impl From<CreateGraphNode> for Statement {
+    fn from(statement: CreateGraphNode) -> Statement {
+        Statement::CreateGraphNode(statement)
+    }
+}
+
+/// A `let` statement that declares a new immutable variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct DeclareImmutable {
+    pub variable: Variable,
+    pub value: Expression,
+}
+
+impl From<DeclareImmutable> for Statement {
+    fn from(statement: DeclareImmutable) -> Statement {
+        Statement::DeclareImmutable(statement)
+    }
+}
+
+/// A `var` statement that declares a new mutable variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct DeclareMutable {
+    pub variable: Variable,
+    pub value: Expression,
+}
+
+impl From<DeclareMutable> for Statement {
+    fn from(statement: DeclareMutable) -> Statement {
+        Statement::DeclareMutable(statement)
+    }
+}
+
+/// A `print` statement that prints out some debugging information
+#[derive(Debug, Eq, PartialEq)]
+pub struct Print {
+    pub values: Vec<Expression>,
+}
+
+impl From<Print> for Statement {
+    fn from(statement: Print) -> Statement {
+        Statement::Print(statement)
+    }
+}
+
+/// A `scan` statement that matches regular expressions against a string
+#[derive(Debug, Eq, PartialEq)]
+pub struct Scan {
+    pub value: Expression,
+    pub arms: Vec<ScanArm>,
+}
+
+impl From<Scan> for Statement {
+    fn from(statement: Scan) -> Statement {
+        Statement::Scan(statement)
+    }
+}
+
+/// One arm of a `scan` statement
+#[derive(Debug)]
+pub struct ScanArm {
+    pub regex: Regex,
+    pub statements: Vec<Statement>,
+}
+
+impl Eq for ScanArm {}
+
+impl PartialEq for ScanArm {
+    fn eq(&self, other: &ScanArm) -> bool {
+        self.regex.as_str() == other.regex.as_str() && self.statements == other.statements
+    }
+}
+
+/// A reference to a variable
+#[derive(Debug, Eq, PartialEq)]
+pub enum Variable {
+    Scoped(ScopedVariable),
+    Unscoped(UnscopedVariable),
+}
+
+/// A reference to a scoped variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct ScopedVariable {
+    pub scope: Box<Expression>,
+    pub name: Identifier,
+}
+
+impl From<ScopedVariable> for Variable {
+    fn from(variable: ScopedVariable) -> Variable {
+        Variable::Scoped(variable)
+    }
+}
+
+/// A reference to a global or local variable
+#[derive(Debug, Eq, PartialEq)]
+pub struct UnscopedVariable {
+    pub name: Identifier,
+}
+
+impl From<UnscopedVariable> for Variable {
+    fn from(variable: UnscopedVariable) -> Variable {
+        Variable::Unscoped(variable)
+    }
+}
+
+impl From<Identifier> for Variable {
+    fn from(name: Identifier) -> Variable {
+        UnscopedVariable { name }.into()
+    }
+}
+
+/// An expression that can appear in a graph DSL file
+#[derive(Debug, Eq, PartialEq)]
+pub enum Expression {
+    // Literals
+    FalseLiteral,
+    NullLiteral,
```

Comment:
What are the semantics of null?

Reply:
There don't really have semantics in the graph DSL itself, but they're there because you might want them as attributes that you attach to node and edges.  (I'm basically aiming for the JSON data model there, since that's the output format we'll use in an as-yet-unwritten PR)